File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int helper(int n, int m,vector<vector<int>>grid,vector<vector<int>>& dp){
        if(n>=0 && m>=0 && grid[n][m]==1){
            return 0;
        }
        if(m==0 && n==0){
            return 1;
        }
        if(m<0 || n<0){
            return 0;
        }
        if(dp[n][m]!=-1){
            return dp[n][m];
        }
        return dp[n][m]=helper(n-1,m,grid,dp)+helper(n,m-1,grid,dp);
    }
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int n=obstacleGrid.size();
        int m=obstacleGrid[0].size();
        vector<vector<int>>dp(n,vector<int>(m,-1));
        return helper(n-1,m-1,obstacleGrid,dp);

    }
};

File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int helper(int i,int j,int n,int m, int[][] obstacleGrid,int[][] dp){
        if(i>=n || j>=m){
            return 0;
        }

if (obstacleGrid[i][j] == 1) return 0;
        if(i==n-1 && j==m-1){
            return 1;
        }

        

        if(dp[i][j]!=-1){
            return dp[i][j];
        }

        return dp[i][j]=helper(i+1,j,n,m,obstacleGrid,dp)+helper(i,j+1,n,m,obstacleGrid,dp);
    }

    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int n=obstacleGrid.length;
        int m=obstacleGrid[0].length;

        int[][] dp=new int[n][m];
        if (obstacleGrid[0][0] == 1) return 0;
        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],-1);
        }
        return helper(0,0,n,m,obstacleGrid,dp);
    }

}